File: BT Training kiến thức về xử lý dữ liệu/huannv_Databinding/app/src/main/java/com/example/base_project/base/BaseFragment.java
package com.example.base_project.base;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import butterknife.ButterKnife;

public abstract class BaseFragment extends Fragment {
    protected Activity activity;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activity = getActivity();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(getLayoutId(), container, false);
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initFragment();
    }
    protected abstract int getLayoutId();

    protected abstract void initFragment();

    protected <VM extends BaseViewModel>  VM getViewModel(Class<? extends BaseViewModel> viewModelType) {
        ViewModelProvider.Factory factory = new ViewModelProvider.NewInstanceFactory();
        VM vm = (VM) new ViewModelProvider(this, factory).get(viewModelType);
        vm.setActivity(activity);
        return vm;
    }


    protected void onBackFragment() {
        activity.onBackPressed();
    }

}

File: Trainning_Kotlin/app/src/main/java/com/example/myapplication/base/BaseFragment.kt
package com.example.myapplication.base

import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import butterknife.ButterKnife

abstract class BaseFragment : Fragment() {
    protected lateinit var activity : Activity

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activity = getActivity()!!
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        var rootView : View = inflater.inflate(getLayoutId(), container, false)
        ButterKnife.bind(this, rootView)
        return rootView
    }

    abstract fun getLayoutId(): Int

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initFragment()
    }

    abstract fun initFragment()

    protected fun onBackFragment() {
        activity.onBackPressed()
    }

    protected fun <VM : BaseViewModel?> getViewModel(viewModelType: Class<in BaseViewModel?>?): VM {
        val factory: ViewModelProvider.Factory = NewInstanceFactory()
        val vm = ViewModelProvider(this, factory).get(viewModelType) as VM
        vm!!.setActivity(activity)
        return vm
    }
}